### PR TITLE
Open Paths Directly from URL

### PR DESCRIPTION
**Pull request by Dala0**
## What does this change?

Open paths directly from the URL rather than querying pluto-core with a project id. to get the correct path.

The paths are checked against two white lists before an attempt is made at opening the file. One white list is for the start of the path and the other the file extension.

A white list is also now used to check the start of asset folder paths before attempting to open them.

The user can edit the white lists in the Preferences window.

Please note, this should not be merged before the required changes to pluto-core have been made as it will render the software useless without them. These changes can be found in the merge request at the following URL: -

https://gitlab.com/codmill/customer-projects/guardian/pluto-core/-/merge_requests/150

## How to test

Install the required changes to pluto-core on a server. Run the new build of PlutoHelperAgent. Add the required paths and extensions for the file you want to try and open. Press the open button in pluto-core. The file should be passed to the open script you have set up. If not, you should see an alert window explaining what went wrong.

## How can we measure success?

Project files can now be opened directly from the URL which does not require the user to be logged into Pluto in PlutoHelperAgent.

## Have we considered potential risks?

Yes. The white lists should stop most if not all attempts at running malicious software.

## Images

image

### CODE REVIEW

File: PlutoHelperAgent/AppDelegate.m [hunk 1]
```diff
@@ -73,92 +47,67 @@ - (void)basicErrorMessage:(NSString *)msg informativeText:(NSString *)informativ
...
+            NSAlert *alert = [[NSAlert alloc] init];
+            [alert setMessageText:@"Setup problem"];
+            [alert setInformativeText:@"Your mac does not appear to be set up correctly, no helper script is configured. Please contact multimediatech@theguardian.com."];
+            [alert setAlertStyle:NSWarningAlertStyle];
+            [alert runModal];
+        });
+        return;
+    }
+
+    NSString *pathToUse = [NSString stringWithFormat: @"%@", projectPath];
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: `pathToUse` seems redundant here - we can just use `projectPath` in its place now

File: PlutoHelperAgent/AppDelegate.m [hunk 2]
```diff
@@ -73,92 +47,67 @@ - (void)basicErrorMessage:(NSString *)msg informativeText:(NSString *)informativ
...
+                 ];
+                [alert setAlertStyle:NSWarningAlertStyle];
+                [alert runModal];
+            });
+        }
+        if([finishedTask terminationReason]!=NSTaskTerminationReasonExit){
+            dispatch_async(dispatch_get_main_queue(), ^{
+                NSAlert *alert = [[NSAlert alloc] init];
+                [alert setInformativeText:
+                 [NSString stringWithFormat:@"Open script failed because the open script terminated unexpectedly.\n%@\n%@\nPlease contact multimediatech@theguardian.com and send a copy of this message",
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: Not directly related, but please could we update to say "...and send a copy of this message along with the asset tag number of this computer"

File: PlutoHelperAgent/AppDelegate.m [hunk 3]
```diff
@@ -222,56 +171,71 @@ -    (void)getUrl:(NSAppleEventDescriptor *)event
             return;
         }
         
-        if (folderToOpen && isDir) {
+        NSArray *assetWhiteListArray = [[NSUserDefaults standardUserDefaults] arrayForKey:@"asset_list"];
+        BOOL pathGood = false;
+        for (id item in assetWhiteListArray) {
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: `item` is `NSDictionary *` isn't it? `id` will disregard the typing which can be a Bad Thing

File: PlutoHelperAgent/AppDelegate.m [hunk 4]
```diff
@@ -222,56 +171,71 @@ -    (void)getUrl:(NSAppleEventDescriptor *)event
             return;
         }
         
-        if (folderToOpen && isDir) {
+        NSArray *assetWhiteListArray = [[NSUserDefaults standardUserDefaults] arrayForKey:@"asset_list"];
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: I would separate all this into its own method, called something like `checkIsInWhitelist`:

```objective-c
- (bool) checkIsInWhitelist:(NSString *)thing whiteListName:(NSString *)whiteListName
```
which would load the `arrayForKey` from `whiteListName` and check `thing` for having the right prefix.
It should return `true` _immediately_ that it finds a match (no point in checkthing the others) or `false` if we fall out of the loop.

You can then call it at line 174 like:
```objective-c
[ self checkIsInWhitelist:folderToOpen whiteListName:@"asset_list];
```

Which is much more self explanatory

Comment (fredex42) [reply PlutoHelperAgent/AppDelegate.m]: Please also then write some tests for it - correct true/false values, empty inputs, empty outputs etc. to try to catch any ways that an attacker (or user!) might misconfigure it

File: PlutoHelperAgent/AppDelegate.m [hunk 5]
```diff
@@ -222,56 +171,71 @@ -    (void)getUrl:(NSAppleEventDescriptor *)event
...
+        BOOL pathGood = false;
+        for (id item in assetWhiteListArray) {
+            if ([item[@"path"] isNotEqualTo:@""]) {
+                if ([folderToOpen hasPrefix:item[@"path"]]) {
+                    pathGood = true;
+                }
+            }
+        }
+        
+        if (!pathGood) {
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m:176]: Once you've done the above, this can be changed to simply:

```objective-c
if([ self checkIsInWhitelist:folderToOpen whiteListName:@"asset_list]) {
  //actually perform the action
} else {
  //the code that was in if(!pathGood)
}
```

Which again is nicer and simpler to read

File: PlutoHelperAgent/AppDelegate.m [hunk 6]
```diff
@@ -222,56 +171,71 @@ -    (void)getUrl:(NSAppleEventDescriptor *)event
...
         }
         
     } else if ([action isEqualToString:@"openproject"]){
+        NSString *projectPath = [parts objectAtIndex:2];
         
-        NSString *projectid = [parts objectAtIndex:2];
-        
-    
-        [ProjectLockerAndKeychainFunctions check_logged_in:^(enum ReturnValues connectionStatus) {
+        NSArray *pathsWhiteListArray = [[NSUserDefaults standardUserDefaults] arrayForKey:@"paths_list"];
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: See above, this should be replaced with another call to `checkIsInWhitelist`, using `@"paths_list"` for the `whiteListName`

File: PlutoHelperAgent/AppDelegate.m [hunk 7]
```diff
@@ -222,56 +171,71 @@ -    (void)getUrl:(NSAppleEventDescriptor *)event
...
-                    }
-                    [self showError:@"Permission Denied" informativeText:@"Please check you have entered your username and password correctly." showPrefs:YES];
-                    break;
-                case SERVER_ERROR:
-                    [self showError:@"Server Error" informativeText:@"An error occurred on the server. Please report this error to multimediatech@theguardian.com" showPrefs:NO];
-                    break;
-                default:
-                    [self showError:@"Unknown Error" informativeText:@"An unknown error occurred. Please report this error to multimediatech@theguardian.com" showPrefs:NO];
-                    break;
+        NSArray *extensionsWhiteListArray = [[NSUserDefaults standardUserDefaults] arrayForKey:@"extensions_list"];
```
Comment (fredex42) [inline PlutoHelperAgent/AppDelegate.m]: See above, this should be replaced with another call to `checkIsInWhitelist`, using `@"paths_list"` for the `whiteListName`

**Review by fredex42 — CHANGES_REQUESTED** [review]
Needs to be a bit more DRY - see comments inline

File: PlutoHelperAgent/AppDelegate.m (hunk 4 above)
Comment (fredex42) [reply PlutoHelperAgent/AppDelegate.m]: Thinking about it, might be easier to put it in a separate object (.m / .h file) and declare it as a static method, for the purposes of testing:

```objective-c
+ (bool) checkIsInWhitelist:(NSString *)thing whiteListName:(NSString *)whiteListName
```

Then you can simply call it as:

```objective-c
[MyObjectName checkIsInWhitelist:folderToOpen whiteListName:@"asset_list"]
```

without needing to `alloc` or `init` `MyObjectName`

File: PlutoHelperAgent/AppDelegate.m (hunk 1 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Done.

File: PlutoHelperAgent/AppDelegate.m (hunk 2 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Done.

File: PlutoHelperAgent/AppDelegate.m (hunk 3 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Yes, or at least it should be that or an NSMutableDictionary.

File: PlutoHelperAgent/WhiteListProcessor.m [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+#import "WhiteListProcessor.h"
+
+@implementation WhiteListProcessor
+
++ (bool)checkIsInWhitelist:(NSString *)thing whiteListName:(NSString *)whiteListName prefix:(BOOL)prefix
+{
+    NSArray *whiteListArray = [[NSUserDefaults standardUserDefaults] arrayForKey:whiteListName];
+    for (id item in whiteListArray) {
+        if ([item[@"string"] isNotEqualTo:@""]) {
```
Comment (fredex42) [inline PlutoHelperAgent/WhiteListProcessor.m]: personally I would use

```objective-c
if ([item[@"string"] isEqualTo:@""]) continue;
```

to avoid extra indent levels which are always confusing at the end of a function

Comment (Dala0) [reply PlutoHelperAgent/WhiteListProcessor.m]: Done.

File: PlutoHelperAgent/AppDelegate.m (hunk 4 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Done.

File: PlutoHelperAgent/AppDelegate.m (hunk 5 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m:176]: I have done this differently as pathGood is accessed more than once.

File: PlutoHelperAgent/AppDelegate.m (hunk 6 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Done.

File: PlutoHelperAgent/AppDelegate.m (hunk 7 above)
Comment (Dala0) [reply PlutoHelperAgent/AppDelegate.m]: Done.

File: PlutoHelperAgent/CCTestingUserDefaults.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
+@import Foundation;
+
```
Comment (fredex42) [inline PlutoHelperAgent/CCTestingUserDefaults.h:7]: Please include a comment giving the URL that we obtained this code from and the reason why we are using it

File: PlutoHelperAgent/AppDelegate.m (hunk 5 above)
Comment (fredex42) [reply PlutoHelperAgent/AppDelegate.m:176]: This could still be refactored so you don't need to, it's not particularly neat, but it's not such a big deal

**Review by fredex42 — CHANGES_REQUESTED** [review]

File: PlutoHelperAgent/CCTestingUserDefaults.h (hunk 9 above)
Comment (Dala0) [reply PlutoHelperAgent/CCTestingUserDefaults.h:7]: Done.
